engine/src/main/java/org/destinationsol/ui/nui/NUIManager.java

```diff
@@ -0,0 +1,321 @@
+import org.terasology.input.device.MouseAction;
+import org.terasology.input.device.MouseDevice;
+import org.terasology.nui.FocusManager;
+import org.terasology.nui.FocusManagerImpl;
+import org.terasology.nui.TabbingManager;
+import org.terasology.nui.UITextureRegion;
+import org.terasology.nui.backends.libgdx.LibGDXCanvasRenderer;
+import org.terasology.nui.backends.libgdx.LibGDXKeyboardDevice;
+import org.terasology.nui.backends.libgdx.LibGDXMouseDevice;
+import org.terasology.nui.backends.libgdx.NUIInputProcessor;
+import org.terasology.nui.canvas.CanvasImpl;
+import org.terasology.nui.events.NUIKeyEvent;
+import org.terasology.nui.events.NUIMouseButtonEvent;
+import org.terasology.nui.events.NUIMouseWheelEvent;
+import org.terasology.nui.skin.UISkin;
+import org.terasology.nui.widgets.UIButton;
+import org.terasology.nui.widgets.UIText;
+
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedList;
+
+/**
+ *  The NUI Manager is responsible for the initialisation and interaction between the NUI library and the game.
+ *  It manages the rendering and update cycles of NUI widgets, which are contained in {@link NUIScreenLayer}
+ *  containers. Each NUIScreenLayer manages its own UI-specific logic and rendering.
+ */
+public class NUIManager {
+    /**
+     * This LibGDX renderer used for NUI. It shares a SpriteBatch with the main game, although not a ShapeRenderer.
+     */
+    private LibGDXCanvasRenderer canvasRenderer;
+    /**
+     * The game's canvas, which is used for all NUI rendering operations. See also {@link NUIManager#canvasRenderer}.
+     */
+    private CanvasImpl canvas;
+    /**
+     * A blank white texture, used by-default for the text cursor.
+     */
+    private UITextureRegion whiteTexture;
+    /**
+     * The NUI mouse device. Receives input directly form LibGDX, independently of the game.
+     */
+    private MouseDevice mouse;
+    /**
+     * The NUI keyboard device. Receives input directly from LibHDX, independently of the game.
+     */
+    private KeyboardDevice keyboard;
+    /**
+     * This allows NUI to determine which widgets are in-focus at the moment, which is used primarily for tabbing.
+     */
+    private FocusManager focusManager;
+    /**
+     * The default UI skin used by all widgets.
+     */
+    private UISkin skin;
+
+    /**
+     * The UI stack. The elements are rendered from most recently added to least recent, so a stack-like structure
+     * was used.
+     */
+    private Deque<NUIScreenLayer> uiScreens = new LinkedList<>();
+
+    private static final String WHITE_TEXTURE_URN = "engine:uiWhiteTex";
+    private static final String DEFAULT_SKIN_URN = "engine:default";
+    private static final String BUTTON_CLICK_URN = "engine:uiHover";
+    /**
+     * The value 0.9 was found from {@link org.destinationsol.ui.SolInputManager#playClick}, so it was copied here to
+     * retain the same click sound as the built-in UI.
+     */
```

Comment:
Thank you for this comment. There are far too many magic numbers in the codebase :roll_eyes: